packages/medusa/src/workflows/admin/create-products/definition.ts

```diff
@@ -0,0 +1,421 @@
+  next: {
+    action: CreateProductsWorkflowActions.prepare,
+    noCompensation: true,
+    next: {
+      action: CreateProductsWorkflowActions.createProducts,
+      next: [
+        {
+          action: CreateProductsWorkflowActions.attachShippingProfile,
+          saveResponse: false,
+        },
+        {
+          action: CreateProductsWorkflowActions.attachToSalesChannel,
+          saveResponse: false,
+        },
+        {
+          action: CreateProductsWorkflowActions.createPrices,
+          next: {
+            action: CreateProductsWorkflowActions.createInventoryItems,
+            next: {
+              action: CreateProductsWorkflowActions.attachInventoryItems,
+              next: {
+                action: CreateProductsWorkflowActions.result,
+                noCompensation: true,
+              },
+            },
+          },
+        },
+      ],
+    },
+  },
+}
+
+const shouldSkipInventoryStep = (
+  container: MedusaContainer,
+  stepName: string
+) => {
+  const inventoryService = container.resolve(
+    "inventoryService"
+  ) as IInventoryService
+  if (!inventoryService) {
+    const logger = container.resolve("logger")
+    logger.warn(
+      `Inventory service not found. You should install the @medusajs/inventory package to use inventory. The '${stepName}' will be skipped.`
+    )
+    return true
+  }
+
+  return false
+}
+
+export function transactionHandler(
+  dependencies: InjectedDependencies
+): TransactionStepHandler {
+  const { manager, container } = dependencies
+
+  const command = {
+    [CreateProductsWorkflowActions.prepare]: {
+      [TransactionHandlerType.INVOKE]: async (
+        data: CreateProductsWorkflowInputData
+      ) => {
+        return await prepareCreateProductsData({
+          container,
+          manager,
+          data,
+        })
+      },
+    },
+
+    [CreateProductsWorkflowActions.createProducts]: {
+      [TransactionHandlerType.INVOKE]: async (
+        data: CreateProductsData
+      ): Promise<ProductTypes.ProductDTO[]> => {
+        return await createProducts({
+          container,
+          data,
+        })
+      },
+      [TransactionHandlerType.COMPENSATE]: async (
+        data: CreateProductsWorkflowInputData,
+        { invoke }
+      ) => {
+        const products = invoke[
+          CreateProductsWorkflowActions.createProducts
+        ] as ProductTypes.ProductDTO[]
+
+        if (!products?.length) {
+          return
+        }
+
+        return await removeProducts({
+          container,
+          data: products,
+        })
+      },
+    },
+
+    [CreateProductsWorkflowActions.attachShippingProfile]: {
+      [TransactionHandlerType.INVOKE]: async (
+        data: CreateProductsWorkflowInputData,
+        { invoke }
+      ) => {
+        const products = invoke[
+          CreateProductsWorkflowActions.createProducts
+        ] as ProductTypes.ProductDTO[]
+        const { productsHandleShippingProfileIdMap } = invoke[
```

Comment:
thought(very non-blocking): To me, it feels a little bit messy, that we have to pull the input data this way in every step - and potentially extended. It is not important to think much about right now, as no developers will access this in the near future. But if possible, it would be nice to make this a more intuitive DX

Reply:
yes exactly, but we find a solution with carlos using the pipe pattern that we have introduced, where you can pass some "middlewares" to manipulate the data etc or an helper object where you specify where the data comes from and what alias should it be attached to in the data

Reply:
here is the [pipe usage](https://github.com/medusajs/medusa/pull/4553/files#diff-e76289c6ca602d28d757be088b439678a15d11ea85194028778719f7eaee4357R28) and the data access which the [function](https://github.com/medusajs/medusa/pull/4553/files#diff-09a24ce6eff5810a54705bb53d7346b87338ec4d361e30cb9ee943658b69d6deR10) request

Reply:
Got it, I like the idea. I think it will be easier to tell if this should be changed, as we start to play around with it

Reply:
yes exactly, for now we tried to do our best but without really using it. Once we will have a better idea of our needs and usage etc we will be able to polish all of that if needed and continue to move forward.